patches/server/0971-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,1534 @@
+index 30aec9dff249ae629b22318e52902361a9fa4099..c2bf7bca7531846b402a24d5fa72b353f846cc03 100644
+--- a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
++++ b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
+@@ -187,7 +187,7 @@ public class ItemFrame extends HangingEntity {
+         } else if (!source.is(DamageTypeTags.IS_EXPLOSION) && !this.getItem().isEmpty()) {
+             if (!this.level.isClientSide) {
+                 // CraftBukkit start - fire EntityDamageEvent
+-                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, amount, false) || this.isRemoved()) {
++                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, io.papermc.paper.entity.damageorigin.DamageOriginWrapper.toApi(source), amount, false) || this.isRemoved()) { // Paper - Damage source wrapper
+                     return true;
+                 }
+                 // CraftBukkit end
+diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+index d47b3ac633e7936d30abfda6fc46c2c7412d76fe..1e1eb5b053efc5b4a5b9629107949daa7dd92c2e 100644
+--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
++++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+@@ -347,7 +347,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
+             return true;
+         } else {
+             // CraftBukkit start
+-            if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, amount)) {
++            if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, io.papermc.paper.entity.damageorigin.DamageOriginWrapper.toApi(source), amount)) { // Paper - Damage source wrapper
+                 return false;
+             }
+             // CraftBukkit end
+diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
+index 2b02800666b358159c8ecb63208a14855f90657b..f90ca3efe3ab6725a6027f2526e6555bd14de167 100644
+--- a/src/main/java/net/minecraft/world/entity/player/Player.java
++++ b/src/main/java/net/minecraft/world/entity/player/Player.java
+@@ -1017,13 +1017,13 @@ public abstract class Player extends LivingEntity {
+     }
+ 
+     @Override
+-    protected void hurtCurrentlyUsedShield(float amount) {
++    protected void hurtCurrentlyUsedShield(DamageSource source, float amount) {
+         if (this.useItem.is(Items.SHIELD)) {
+             if (!this.level.isClientSide) {
+                 this.awardStat(Stats.ITEM_USED.get(this.useItem.getItem()));
+             }
+ 
+-            if (amount >= 3.0F) {
++            if (source.hurtShield.toBooleanOrElse(amount >= 3.0F)) { // Paper
```

Comment:
Hmm, so is this custom behavior that we are introducing?

Reply:
shuutt there's nothinggg..

Reply:
Take a look at LivingEntity#isDamageSourceBlocked

I think that instead this could be handled by adding our custom damage source to the BYPASSES_SHIELD tag.
Now, this is covered by another PR.

Reply:
This is not the equivalent to the BYPASSES_SHIELD tag. The bypass only check if the damage is blocked or not but not if it is damaged or not. With this option you can create a damage that is blocked but doesn't hurt the shield even for damage >= 3